### ISSUE TITLE
Alldebrid puts files in silly places sometimes, causing rdt-client to be unable to find them

### ISSUE THREAD

**Issue by Cucumberrbob** [rogerfar/rdt-client#648]
**What version are you using?**
docker 2541993a2fe6

**What OS are you running?**
NixOS

**Are you using Docker or as a service?**
Docker

**Which debrid provider are you using?**
Alldebrid

**Which downloader are you using?**
Symlink

**Please attach a log file here with the log setting set to debug**
See below for excerpts where relevant

Edit: [Note] AD changed how single file torrents are stored in the webDAV drive on 2025-01-07T12:30:00Z ish - many comments about single-file torrents are no longer true

Variables in `<>` referenced below are from the `data.magnets` field in a response from a `/api/v4/magnet/status` request.

Alldebrid is doing some weird stuff with their webdav. I've identified a few cases where `rdt-client` won't pick up the files. Below  are the cases I've identified 
- Single file torrent whose name does not end in a file extension, but the file inside does: The file is available at `/mnt-path/magnets/<filename>` (notably, not `links[0].filename`).
- Torrents  containing a subdirectory. Say there was a file in this subdirectory, then it would be available at `/mnt-path/magnets/<filename>/<links[I].files[0].n]>/<links[0].files[0].e[0].n>`.  

As an example of the first case: say `<filename> = "A"` and `<links[0]> = {"filename": "B.mkv", "files": [{"n": "B.mkv"}]}` Then the file is at `/mnt-path/magnets/A` (no extension). Yes this is really an mkv file despite having no extension. . The logs for the  files checked are as follows
```
[DBG] Searching /mnt/rclone/magnets for B.mkv (attempt #3)...
[DBG] Searching /mnt/rclone/magnets/A/B.mkv...
[DBG] Searching /mnt/rclone/magnets/B.mkv/B.mkv...
[DBG] Searching /mnt/rclone/magnets/B/B.mkv...
[DBG] Searching /mnt/rclone/magnets/B.mkv...
```

For the second case, say `<filename> = "A"`, and there are files `C.mkv` and `D.mkv` in the subdirectory B. Then the files are at `/mnt-path/magnets/A/B/C.mkv` and `/mnt-path/magnets/A/B/D.mkv`. The logs for the files checked are the same as above.

The `data.magnets`  in the response from `/api/v4/magnet/status` looks something like
```json
{
  "filename": "A",
  "links": [
    {
      "filename": "C.mkv",
      "files": [{
        "n": "B",
        "e": [{
          "n": "C.mkv"
        }]
      }]
    },
   {
      "filename": "D.mkv",
      "files": [{
        "n": "B",
        "e": [{
          "n": "D.mkv"
        }]
      }]
    }
  ]
}
```
which, when constructed as above , does produce the correct filename. This construction also works for torrents which do not contain subdirectories, just that `<links[0].files[0]>` is `{ "n": "B.mkv"}`, with no `e`, so the correct filename is `/mnt-path/magnets/<filename>/<links[I].files[0].n]>`. 
However, this construction fails for single file torrents, in a similar way to the current setup: most single file torrents are named the same as the file they contain, so it's fine most of the time, but if the `filename ` is not the same as `links[0].n`, the construction fails. . For single file torrents, `/mnt-path/magnets/<filename>` is always the correct path.

Effectively, I think the logic is that if there is one file in the torrent (even if it is in a subdirectory), it appears at `/mnt-path/magnets/<filename>`, and if there is more than one file, each file's path is `/mnt-path/magnets/<filename>/<all the `"n"`s of its parent>/<its "n">`

**Comment by Cucumberrbob** [comment]
@rogerfar that PR does not fully close this issue - the single file torrents are still a problem. I can open a new issue if you'd rather they be separate.

**Comment by rogerfar** [comment]
It must've closed this automatically when I merged the PR.

@Sculas is this something you could have a look at?

**Comment by Sculas** [comment]
> As an example of the first case: say `<filename> = "A"` and `<links[0]> = {"filename": "B.mkv", "files": [{"n": "B.mkv"}]}` Then the file is at `/mnt-path/magnets/A` (no extension). Yes this is really an mkv file despite having no extension. The logs for the files checked are as follows
> 
> ```
> [DBG] Searching /mnt/rclone/magnets for B.mkv (attempt #3)...
> [DBG] Searching /mnt/rclone/magnets/A/B.mkv...
> [DBG] Searching /mnt/rclone/magnets/B.mkv/B.mkv...
> [DBG] Searching /mnt/rclone/magnets/B/B.mkv...
> [DBG] Searching /mnt/rclone/magnets/B.mkv...
> ```

If I understand you correctly, `A` is the name of the torrent, and `B.mkv` is the only file in it.
This means the file should be at `/mnt/rclone/magnets/B.mkv`. I've confirmed this by downloading a torrent from Nyaa that had a different name than the single file in it. Connecting to the WebDAV server via WinSCP showed the file with a different name in the `magnets` directory, as I expected. Requesting the torrent via Alldebrid's API shows a single file with a different name. RDT does the same, it sees the file and then sends that directly to the downloader. This should all just work as expected?

@Cucumberrbob, is this still the behavior after my PR? I don't see a reason at all why RDT would check the `A` directory here.
RDT knows the **exact** path each file will be at (by asking Alldebrid beforehand) in the rclone mount, so it technically shouldn't even need to "search" for it. I don't have an rclone mount setup, so if you could check for me that would be really appreciated :)

**Comment by Sculas** [comment]
Wait but, so now it's `/mnt/rclone/magnets/<torrent name>/<torrent name>/<file name>`, right?
And why does that work?! It should be `/mnt/rclone/magnets/<file name>` for all I know.
Do you have a torrent or magnet I can reproduce this with? If needed, you can encode it in base64 so it doesn't get indexed.

**Comment by Sculas** [comment]
Oh, that's interesting. A torrent only allows a single file or a directory with multiple files. It seems AllDebrid makes you blindly follow that assumption because the API doesn't show the root folder if it has one.

After much reading, I think I've been able to find the root cause:

https://github.com/rogerfar/rdt-client/blob/79164d54ca5dfe679113c80a3e86b8e8846a44da/server/RdtClient.Service/Helpers/DownloadHelper.cs#L59-L99

This function always adds the torrent filename, which is actually the **root directory** of the torrent, to the file path.
This is evident when using a downloader like aria2c because the file gets placed at `<download path>/<torrent filename>/<the file that's in the torrent>`. So the fix I'm thinking is that we check if the torrent has multiple files, and if so don't add the torrent filename to the file path.

@rogerfar Do you know if this behavior also applies to RealDebrid? Does it place single-file torrents in a directory or not?

**Comment by rogerfar** [comment]
I'm testing a bit and I'm not sure if this is what we're trying to accomplish.

I attached a couple of test torrents I uploaded myself. When I download them with qBittorrent, I get these results with `tree /F`:

```
C:.
│   TestFileNameMismatch.txt
│   TestSingle.txt
│
├───TestMultiple
│       TestMultiple.txt
│       TestMultiple2.txt
│
└───TestSingleInSub
    └───TestSingleInSub
            TestSingleInSub.txt
```

Before your patch I get:
```
C:.
├───TestFileNameMismatch.txt
│       TestFileNameMismatch.txt
│
├───TestMultiple
│       TestMultiple.txt
│       TestMultiple2.txt
│
├───TestSingle.txt
│       TestSingle.txt
│
└───TestSingleInSub.txt
    └───TestSingleInSub
            TestSingleInSub.txt
```

After your patch it's the same.

This isn't really good behaviour anyway as I want to match the output of RDT similar to qBittorrent.

If you run these test torrents in AllDebrid, how do they end up?

[TestTorrents.zip](https://github.com/user-attachments/files/18312090/TestTorrents.zip)

**Comment by Sculas** [comment]
That was indeed my intention as well, to have the same output as qBittorrent. I will test the torrents you've provided tomorrow! (currently night for me)

**Comment by rogerfar** [comment]
I found 1 snag, this is the Webdav drive from Real-Debrid when adding those test torrents:

```
Z:.
├───TestSingleInSub.txt
│       TestSingleInSub.txt
│
├───TestMultiple
│       TestMultiple.rar
│
└───TestSingle.txt
        TestSingle.txt
```

It seems it never places it in the root, even when it's a single file, and it ignores my sub-directory.

**Comment by rogerfar** [comment]
I added some code in my latest commit if you want to do a snifftest. I would like to see the results from AllDebrid and TorBox too to compare.

**Comment by Sculas** [comment]
Okay, so RealDebrid always stores it in a directory. AllDebrid doesn't. That's frustrating since we'll have to handle that.

> Which seems to not matter, the symlink is what matters.

It does matter, though? The issue is that RDT always assumes the files are in a directory, which is the case for RealDebrid, but not for AllDebrid, and then tries to access the file from the WebDAV drive at a place it thinks it exists, but doesn't.

See it like this. Let's say we're downloading `RDT.Client.v2.0.94.exe.torrent`, which contains a single file: `RDT.Client.v2.0.94.exe`
- RealDebrid stores this at: `<webdav>/RDT.Client.v2.0.94.exe/RDT.Client.v2.0.94.exe`
- AllDebrid stores this at: `<webdav>/RDT.Client.v2.0.94.exe`
- RDT will assume this file is at: `<webdav>/RDT.Client.v2.0.94.exe/RDT.Client.v2.0.94.exe`

Now, let's download `RDT.Client.All.Versions.torrent`, which looks like this:
```
C:.
└───RDT.Client.All.Versions
        RDT.Client.v2.0.92.exe
        RDT.Client.v2.0.93.exe
        RDT.Client.v2.0.94.exe
```
Let's use `RDT.Client.v2.0.92.exe` as the file here we're targeting.
- RealDebrid stores this at: `<webdav>/RDT.Client.All.Versions/RDT.Client.v2.0.92.exe`
- AllDebrid stores this at: `<webdav>/RDT.Client.All.Versions/RDT.Client.v2.0.92.exe`
- RDT will assume this file is at: `<webdav>/RDT.Client.All.Versions/RDT.Client.v2.0.92.exe`

> If you run these test torrents in AllDebrid, how do they end up?

~~It seems the torrents are stalled. Are they still being seeded?~~ 2/4 have downloaded now, but the others are still stalled.

**Comment by Sculas** [comment]
With those two torrents downloaded, I currently have this in `/data/downloads`:
```
C:.
│   TestFileNameMismatch.txt
│
└───TestSingleInSub
    └───TestSingleInSub
            TestSingleInSub.txt
```
I don't have the WebDAV drive mounted so I can't show a tree output, but I've checked and it's the same as above.
So, currently, RDT and AllDebrid match in their file structure. So, it seems like we'll have to add a check for RealDebrid, then.

**Comment by Cucumberrbob** [comment]
> If I understand you correctly, `A` is the name of the torrent, and `B.mkv` is the only file in it. This means the file should be at `/mnt/rclone/magnets/B.mkv`. I've confirmed this by downloading a torrent from Nyaa that had a different name than the single file in it. Connecting to the WebDAV server via WinSCP showed the file with a different name in the `magnets` directory, as I expected. Requesting the torrent via Alldebrid's API shows a single file with a different name. RDT does the same, it sees the file and then sends that directly to the downloader. This should all just work as expected?

That's really weird, because I _would_ expect that file to appear at `/mnt/rclone/magnets/A` - in fact [Name Mismatch (Single File).torrent.zip](https://github.com/user-attachments/files/18318746/Name.Mismatch.Single.File.torrent.zip) torrent demonstrates this behaviour: It appears at `/mnt/rclone/magnets/Name\ Mismatch\ \(Single\ File\)`, not `/mnt/rclone/magnets/single-file.txt`. I created that torrent using `mktorrent Name\ Mismatch\ \(Single\ File\)` where `Name\ Mismatch\ \(Single\ File\)` is a directory with `single-file.txt` inside.
Here's the `/v4/magnets/status` response with irrelevant fields omitted.
```json
{
    "data": {
        "magnets": {
            "filename": "Name Mismatch (Single File)",
            "links": [
                {
                    "filename": "single-file.txt",
                    "files": [
                        {
                            "n": "Name Mismatch (Single File)",
                            "e": [
                                {
                                    "n": "single-file.txt",
                                }
                            ]
                        }
                    ],
                }
            ],
        }
    },
}
```
I can confirm this is still not found by the latest version (`docker build` run on `d5ee5d2`)

I have at least one single-file torrent that is even weirder: its `/api/v4/status` looks like:
```json
{
    "data": {
        "magnets": {
            "filename": "A",
            "links": [
                {
                    "filename": "B.mkv",
                    "files": [
                        {
                            "n": "B.mkv",
                        }
                    ],
                }
            ],
        }
    },
}
```
And is available at `/mnt/rclone/magnets/A`. I'm not sure how this torrent was made - I've been unable to figure out how to recreate this structure with either qBittorrent's torrent creator or mktorrent. But I think it confirms my suspicion that it's `filename`, not the first `n` which is used to make the name of the file that will appear in `/mnt/rclone/magnets`

**Comment by Cucumberrbob** [comment]
Also worth noting that the `TestFileNameMismatch.txt`  torrent doesn't actually demonstrate this mismatch behaviour:
```json
{
    "data": {
        "magnets": {
            "filename": "TestFileNameMismatch.txt", // <-- same as links[0].files[0].n, doesn't demonstrate mismatch
            "links": [
                {
                    "filename": "TestFileNameMismatch.txt",
                    "files": [
                        {
                            "n": "TestFileNameMismatch.txt"
                        }
                    ]
                }
            ]
        }
    }
}
```

**Comment by Sculas** [comment]
AllDebrid's way of placing files works exactly like qBittorrent, with `Save at` set to `/magnets`.
At the root, a torrent must only contain 1 file or directory. If you want multiple files, it needs to be in a directory.

As for the torrent you provided, the WebDAV server reports that the file is at `magnets/Name Mismatch (Single File)/single-file.txt`, is this also what you see? Because this is what I expect. You can verify this by opening the torrent in qBittorrent, and seeing that it's structured like this: `Name Mismatch (Single File)/single-file.txt`. If I now download this file using RDT, it's placed at `/data/downloads/Name Mismatch (Single File)/single-file.txt`. This is again the result I expect to see.

So why would you expect this file to be at `/data/downloads/single-file.txt`, if you're the one that placed the file in a directory to begin with? (if I'm supposed to believe what qBittorrent is telling me about your torrent)

**Comment by Cucumberrbob** [comment]
I think we might be misunderstanding each other. I'm not making any assertions as to where files should end up. I'm saying that when a torrent like `Name Mismatch (Single File)` (which has this mismatch in `/v4/magnets/status`) is added to AD, it will appear at `/mnt/rclone/magnets/<filename>` . Rdt-client probably _should_ create a symlink at /data/downloads/<filename>/<links[0].files[0].e[0].n>`.

But I disagree with your first sentence:
>AllDebrid's way of placing files works exactly like qBittorrent, with Save at set to /magnets.

For `Name Mismatch (Single File)`, when I add it to qBitTorrent, it will be saved at `/qBit-dl/Name\ Mismatch\ \(Single\ File\)/single-file.txt`. But when I add it to AD, it appears at `/mnt/rclone/magnets/Name\ Mismatch\ \(Single\ File\)`

**Comment by Cucumberrbob** [comment]
@bob1321  are you able to test with [this file](https://github.com/user-attachments/files/18318746/Name.Mismatch.Single.File.torrent.zip)? Maybe this is something on my end instead

**Comment by Sculas** [comment]
Oh wait, are we experiencing the same issue then? I've just been working around it since I thought it was a weird issue only I had.
This is what I see if I connect to the WebDAV server via WinSCP:

![image](https://github.com/user-attachments/assets/75e6bb10-3f52-481a-8897-3f096b63dd67)

As you can see, the torrents aren't directories for some reason, but instead something... weird?
I can't open the directories in WinSCP either, I have to right-click them and press "Open", only then does it download the directory to a temp directory and does it properly show the folder.

It sounds wrong that the file is placed at `/mnt/rclone/magnets/Name Mismatch (Single File)`, since it should obviously be at `/mnt/rclone/magnets/Name Mismatch (Single File)/single-file.txt`. Can you check if this is also the case for you @Cucumberrbob?

**Comment by Cucumberrbob** [comment]
Yes! spot on! These files are actually files not directories despite how they're named, `TestSingleInSub` is actually a `.txt` - can you `cat` it (or the windows equivalent of `cat` it)?

**Comment by Sculas** [comment]
> Yes! spot on! These files are actually files not directories despite how they're named, `TestSingleInSub` is actually a `.txt` - can you `cat` it (or the windows equivalent of `cat` it)?

Well, that's the thing. If I double click it to open it in something like Notepad, it gives me an "Access denied" error.
That's when I figured out I needed to click "Open" in WinSCP, which downloads it locally, which seems to "fix" the bug.

**Comment by Sculas** [comment]
```console
❯ ls -lah
drwxr-xr-x 1 root root    0 Jan  6 14:16  .
drwxr-xr-x 4 root root 4.0K Jan  6 14:13  ..
-rw-r--r-- 1 root root   16 Jan  6 13:47 'Name Mismatch (Single File)'
-rw-r--r-- 1 root root   77 Jan  6 11:42  TestFileNameMismatch.txt
-rw-r--r-- 1 root root   86 Jan  6 12:28  TestSingleInSub
-rw-r--r-- 1 root root   36 Jan  6 13:30  TestSingle.txt
```
```console
❯ cat 'Name Mismatch (Single File)'
cat: 'Name Mismatch (Single File)': Input/output error
```
```console
❯ cd 'Name Mismatch (Single File)'
cd: not a directory: Name Mismatch (Single File)
```
What?! This has to be a bug then with AllDebrid's WebDAV server and single-file torrents inside a directory.
And yes, other files work fine. Hell, even the file above works in WinSCP with a workaround.
```console
❯ cat TestSingle.txt
This torrent contains a single file.
```

**Comment by Sculas** [comment]
This is so weird. I seriously don't get it anymore. The last torrent, `TestMultiple.torrent`, has completed now.
qBittorrent shows the following folder structure for it:
![image](https://github.com/user-attachments/assets/1ed5337c-76c7-4ebe-b8d6-8cde39fc95f4)
Yet, AllDebrid stored this at `/mnt/rclone/magnets/TestMultiple/TestMultiple/TestMultiple.txt`.
This seems fine until you look at this. If I open a different torrent in qBittorrent, it shows the following folder structure:
![image](https://github.com/user-attachments/assets/dd3a6966-8809-4fd1-a357-191d381d14a5)
Looks similar to the folder structure above, right? Yeah, AllDebrid stored this in a completely different way:
`/mnt/rclone/magnets/[Tenrai-Sensei] Sword Art Online [BD][1080p][HEVC 10bit x265][Dual Audio]/poster.jpg`

So why, why?!?! does AllDebrid store the first torrent in a subfolder, and not the second one? Why the inconsistency?

**Comment by Cucumberrbob** [comment]
What do the `/v4/magnets/status` responses look like for those two? In my experience, the paths are reconstructable from the response.

**Comment by Sculas** [comment]
I'm using `v4.1` here since `v4` is deprecated and imo a bit harder to read, but that doesn't affect the end result.
I've removed fields that aren't necessary, and the SAO torrent had way too much files, so I culled them as well.

---

<details>
<summary><b>TestMultiple:</b></summary>

Stored as: `/mnt/rclone/magnets/TestMultiple/TestMultiple/TestMultiple.txt`

```json
{
  "status": "success",
  "data": {
    "magnets": {
      "filename": "TestMultiple",
      "size": 74,
      "nbLinks": 2,
      "files": [
        {
          "n": "TestMultiple",
          "e": [
            {
              "n": "TestMultiple2.txt",
              "s": 37,
              "l": "<>"
            },
            {
              "n": "TestMultiple.txt",
              "s": 37,
              "l": "<>"
            }
          ]
        }
      ]
    }
  }
}
```

</details>

---

<details>
<summary><b>Sword Art Online:</b></summary>

Stored as: `/mnt/rclone/magnets/[Tenrai-Sensei] Sword Art Online [BD][1080p][HEVC 10bit x265][Dual Audio]/poster.jpg`

```json
{
  "status": "success",
  "data": {
    "magnets": {
      "filename": "[Tenrai-Sensei] Sword Art Online [BD][1080p][HEVC 10bit x265][Dual Audio]",
      "size": 58880130060,
      "nbLinks": 159,
      "files": [
        {
          "n": "Season 1",
          "e": [
            {
              "n": "Sword Art Online - S01E01 - The World Of Swords.mkv",
              "s": 489401205,
              "l": "<>"
            }
          ]
        },
        {
          "n": "poster.jpg",
          "s": 494325,
          "l": "<>"
        }
      ]
    }
  }
}
```

</details>

---

<details>
<summary><b>TestSingleInSub:</b></summary>

Stored as: `/mnt/rclone/magnets/TestSingleInSub` **(broken, cannot access as file, cannot cd into it)**

```json
{
  "status": "success",
  "data": {
    "magnets": {
      "filename": "TestSingleInSub",
      "size": 86,
      "nbLinks": 1,
      "files": [
        {
          "n": "TestSingleInSub",
          "e": [
            {
              "n": "TestSingleInSub",
              "e": [
                {
                  "n": "TestSingleInSub.txt",
                  "s": 86,
                  "l": "<>"
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

</details>

---

**Comment by Cucumberrbob** [comment]
Yep, these are all constructible from the responses:
### For TestMultiple, our file nodes (object with `n` but no `e`) are as follows:
```json
{
  "n": "TestMultiple2.txt",
  "s": 37,
  "l": "<>"
},
{
  "n": "TestMultiple.txt",
  "s": 37,
  "l": "<>"
}
```

And they have the same parent, whose `n` is `TestMultiple`. The shared parent node has not parent (it's an element of `files`), so  we stop and assemble the filename: `<magnets.filename>/<files[0].n>/<files[0].e[i].n>` with `i=0` for the first file `TestMultiple2.txt` and `I=1` for the second file `TestMultiple.txt`. Which is
```
TestMultiple/TestMultiple/TestMultiple.txt
TestMultiple/TestMultiple/TestMultiple2.txt
```

### For the SAO one, we do similar, but since `files.length == 2`, let's do each one individually:
The first element of `files` 
```json
 {
  "n": "Season 1",
  "e": [
    {
      "n": "Sword Art Online - S01E01 - The World Of Swords.mkv",
      "s": 489401205,
      "l": "<>"
    }
  ]
}
 ```
 
 So again, combining the root `filename` with all the `n`s gives us `<filename>/<files[0].n>/<files[0].e[0].n>` - in this case `
 ```
 [Tenrai-Sensei] Sword Art Online [BD][1080p][HEVC 10bit x265][Dual Audio]/Season 1/Sword Art Online - S01E01 - The World Of Swords.mkv
 ```
 
And
```json
 {
  "n": "poster.jpg",
  "s": 494325,
  "l": "<>"
}
```
Corresponds to `<filename>/<files[1].n>`, which is 
```
 [Tenrai-Sensei] Sword Art Online [BD][1080p][HEVC 10bit x265][Dual Audio]/poster.jpg
```

**Comment by Sculas** [comment]
Yeah. It's confusing then why qBittorrent shows something different. Oh well.
Also, @Cucumberrbob what are your rclone parameters? Since I'm not able to access `Name Mismatch (Single File)'` at all.
Regardless, I think AllDebrid is missing this edge case we explicitly check in RDT, which causes the bug:

https://github.com/rogerfar/rdt-client/blob/ee9e80f7d2d595688b0817a40eeb0a9448653864/server/RdtClient.Service/Helpers/DownloadHelper.cs#L63-L73

**Comment by Cucumberrbob** [comment]
My rclone parameters are 
```
--cache-dir /home/cucumberrbob/.cache/rclone --dir-cache-time 10s --cutoff-mode cautious --vfs-cache-mode full  --buffer-size=0 --read-only --allow-non-empty  --vfs-cache-max-size 120G --vfs-cache-max-age 72h --allow-other --config /run/secrets/rclone.conf
```

>I think I'll have to unfortunately leave it at this though. This is something AllDebrid has to fix on their end if we ever want to see this issue closed unless there's some kind of workaround I'm not aware of at the moment. In either case, this doesn't seem to be an issue in RDT.

I think you're pretty much right - without AD fixing their webDAV some of the single file torrents will be unplayable. It's worth looking a bit more to see if this is consistently predictable based on the `/magnets/status` response,
Maybe just erroring out for single file torrents which have this mismatch is the best option

**Comment by Sculas** [comment]
Oh, it is predictable. If we fall into the block where we set the `torrentPath`, it's guaranteed failure:
https://github.com/rogerfar/rdt-client/blob/ee9e80f7d2d595688b0817a40eeb0a9448653864/server/RdtClient.Service/Helpers/DownloadHelper.cs#L63-L73

---

I'm currently reading the verbose logs of WinSCP to figure out what it's doing. I'll update this message as I go.
This message especially caught my eye:
```
2025-01-06 15:47:31.277 Getting size of directory "/magnets/Name Mismatch (Single File)"
```
WinSCP knows it's a directory...?

**Comment by Cucumberrbob** [comment]
> The filename under magnets is only ever used when length(links) > 1.

I don't think this is true. For all the single file torrents I've posted in here, none of them follow that rule. They are all always at `/mnt/rclone/magnets/<filename>`. Could you try the nyaa torrent I posted in the comment above? That ends up at `/mnt/rclone/magnets/<filename>`, but is a playable mkv

**Comment by Sculas** [comment]
You're right, it is stored that way. So then why does that mkv file work, but `Name Mismatch (Single File)` does not, which is literally the exact same thing?

**Comment by Cucumberrbob** [comment]
They're not quite identical - the `/magnets/status` aren't identical in structure - `Name Mismatch (Single File)` has a directory in it, but that nyaa torrent has no directories - it's just a single file at the root.

**Comment by Sculas** [comment]
You're right, qBittorrent's seemingly inconsistent (or AllDebrid's) file listing got me again. Why can these things just not be consistent with each other...

**Comment by Cucumberrbob** [comment]
>You're right, qBittorrent's seemingly inconsistent (or AllDebrid's) file listing got me again. Why can these things just not be consistent with each other... Now, why is AllDebrid missing the file extension on the file...
Honestly I think it's AD again - the first bit of any .torrent file is pretty much human readable if you `head` it , and I think they're the same for torrents that end up with different structures in AD's webDAV.

>But it might be worth it to make a list of all potential iterations and create test torrents for those, because I also see the same discrepancies with Real-Debrid too. qBittorrent seems to have some naming rules itself too which conflicts with expectation and the output of the debrid providers.

Definitely, but I can't figure out how to create torrents of the working kind - where they appear like [this](https://github.com/rogerfar/rdt-client/issues/648#issuecomment-2573340495) in the `/magnets/status` response, and I've been struggling to find GitHub-safe torrents to link to directly.

But between that nyaa one and the `Name Mismatch (Single File)` torrent I uploaded, I think all cases are covered for single-file torrents

**Comment by Sculas** [comment]
> Sorry it's been night here so haven't fully caught up.

The TLDR pretty much is that this is an AllDebrid issue:

1. Torrents with a directory as root and 2 or more files in them: **works** (confirmed by `TestMultiple.torrent`)
2. Torrents with a directory as root and 1 file that may be N directories deep: **broken** (confirmed by `TestSingleInSub.torrent`)
3. Torrents with a file as root: **works** (confirmed by `TestSingle.txt.torrent`)

`Name Mismatch (Single File).torrent` also confirms case 2.
`TestFileNameMismatch.txt.torrent` is the same as `TestSingle.txt.torrent`.

**Comment by Cucumberrbob** [comment]
>Torrents with a file as root: works (confirmed by TestSingle.txt.torrent)

To clarify, this is only slightly true. This only works because `TestSingle.txt.torrent` has the name `TestSingle.txt`, which is the same as the file's name inside it. If the torrent were called `TestSingle` (no `.txt`), this would break.

>TestFileNameMismatch.txt.torrent is the same as TestSingle.txt.torrent.
This is only true because likewise, `TestFilefnameMismatch.txt.torrent` has the same name as the file inside it. Despite its name, it's not actually a mismatch according to `//magnets/status`

**Comment by Cucumberrbob** [comment]
To sum this all up into something actionable:

The logic should be as follows:
```
if the torrent has multiple files then
    Each file's path  is `/mnt-path/<data.magnets.filename>/<all the "n"s>`
else
    // ↓ equivalent to if <data.magnets.links[0].files[0].e> exists
    if the file is inside a directory then
        Error out - these types of torrents confuse AD's webDAV and they aren't usable
    else
      We can use this file! it will appear at `/mnt-path/<data.magnets.filename>`, 
      but we should ensure that it ends up being renamed to `<data.magnets.links[0].files[0].n>`
      (which in my experience has always been the same as `<data.magnets.links[0].filename>`
```

If it's not clear what I mean by "all the 'n's ", I mean all the `"n"` attributes of the ancestor nodes of the file node joined with `/`. A file node is an object that has an `"n"` but has no `"e"`, and is in its parent's `"e"` array (this is what I mean by the parent/child relationship). See [this](https://github.com/rogerfar/rdt-client/issues/648#issuecomment-2573198483) comment for a breakdown of how we construct the file path from the `/magnets/status` response

@Sculas can you spot an issue with this logic?

**Comment by Sculas** [comment]
That looks correct! I'm not sure what's the best way to report this to the Alldebrid developers. Just asking them to test the Nyaa torrent you sent earlier @Cucumberrbob is enough to show something's definitely wrong (missing file extension and broken content/mime type). And hopefully, with that, we can also tell them about the issue with the broken directories (which may be harder to figure out). Their GitHub profile hasn't seen activity in quite some time, unfortunately. I'll send an email to their support.

I may otherwise consider forking [rclone_rd](https://github.com/itsToggle/rclone_rd) and porting it to Alldebrid if I also encounter this issue myself. I'm not using the rclone mount setup right now, but I'm working on switching to it. I'll have to see though, since I've already spent *hours* trying to fix this.

**Comment by Sculas** [comment]
I've sent an email to their support mail (`support [at] alldebrid [dot] com`). Time to wait for their response now.
Also, I'm not sure who sent me an email 2 days ago with a test torrent, but sorry I didn't see it. It went into my spam folder 😅

**Comment by Cucumberrbob** [comment]
Just got a response from AD support - they've said they've found the issue and "will fix it most likely tomorrow". Let's see what we end up with then.
In the meantime, the multi-file torrents are still an issue in the latest version: `TestMultiple.torrent` still is not handled properly. I got something working based on @Sculas' most recent PR. The multi-file part is [here](https://github.com/Sculas/rdt-client/pull/1/files#diff-36327f02bb834221119be3ac5e5a45d2c79653d14fa314c9fe9b18d51764bea5) - which shouldn't actually need the commit from @Sculas' most recent PR.  I'm not sure if adding *another* `potentialFilePath` is a good idea though

**Comment by Sculas** [comment]
Oh wow, that's a quick response! I guess you used the website? I'll take a look at your PR and continue there :)

**Comment by Cucumberrbob** [comment]
Yeah! I won't lie I'm impressed. I did use the contact form on their website.

With regards to the PR, given that the single-file torrent situation will likely change in the near future, only the changes in `DownloadHelper.cs` are actually useful now.

**Comment by Sculas** [comment]
Yeah, I was about to mention that in the PR, but we may want to wait with my PR until we know it's fixed. Otherwise, we push out a release now that works around a part of the bug, but then tomorrow it's fixed for example, and we have to fix it again to remove the workaround. My PR as it is *now* should work just fine afaik, given that the WebDAV bug is fixed in Alldebrid.

**Comment by Cucumberrbob** [comment]
Just heard back from AD (they're insanely quick). As I understand it, the logic is now as follows:
All files are at `/mnt-path/magnets/<filename>/<all the "n"s>` unless it's a single-file-at-the-root torrent, in which case it's `/mnt-path/magnets/<files[0].n>`.

<details>
  <summary><code>ls</code>ing the relevant paths:</summary>
  
  ```
  > eza --tree /mnt/rclone/magnets/{Test*,Name\ Mismatch\ \(Single\ File\)}
  /mnt/rclone/magnets/'Name Mismatch (Single File)'
  └──  single-file.txt
  /mnt/rclone/magnets/TestFileNameMismatch.txt
  /mnt/rclone/magnets/TestMultiple
  └── TestMultiple
      ├── TestMultiple.txt
      └── TestMultiple2.txt
  /mnt/rclone/magnets/TestSingle.txt
  /mnt/rclone/magnets/TestSingleInSub
  └── TestSingleInSub
      └── TestSingleInSub.txt
  ```
</details>

**Comment by Sculas** [comment]
I've been checking occasionally and am happy to report that it's just been fixed! I can finally access the directories again.

**Comment by Cucumberrbob** [comment]
On your PR (ee9e80f) with my changes to the Dockerfile made, a few are still broken:
<details>
  <summary>TestSingleInSub [not working]</summary>

  ```
  [DBG] Starting symlink resolving of [url redacted]/TestSingleInSub.txt, writing to path: TestSingleInSub/TestSingleInSub/TestSingleInSub.txt
  [DBG] Searching /mnt/rclone/magnets for TestSingleInSub.txt (attempt #0)...
  [DBG] Searching /mnt/rclone/magnets/TestSingleInSub/TestSingleInSub.txt...
  [DBG] Searching /mnt/rclone/magnets/TestSingleInSub.txt/TestSingleInSub.txt...
  [DBG] Searching /mnt/rclone/magnets/TestSingleInSub.txt...
  ```
</details>

<details>
  <summary>TestSingle.txt [not working]</summary>

  ````
  [DBG] Starting symlink resolving of [url redacted]/TestSingle.txt, writing to path: TestSingle.txt
  [DBG] Searching /mnt/rclone/magnets for TestSingle.txt (attempt #0)...
  [DBG] Searching /mnt/rclone/magnets/magnets/TestSingle.txt...
  [DBG] Searching /mnt/rclone/magnets/rclone/TestSingle.txt...
  [DBG] Searching /mnt/rclone/magnets/mnt/TestSingle.txt...
  [DBG] Searching /mnt/rclone/magnets/TestSingle.txt/TestSingle.txt...
  [DBG] Searching /mnt/rclone/magnets/TestSingle/TestSingle.txt...
  [DBG] Searching /mnt/rclone/magnets/TestSingle.txt...
  ````
</details>

**Comment by Sculas** [comment]
I suppose that could work, but since the path in `writing to path <path>` is always correct anyway, why not skip the whole search and just get the file directly, since we know it's there anyway? We know the exact behavior of Alldebrid placing the torrent files.

**Comment by Sculas** [comment]
I've fixed the symlink search by adding the `searchPath` to the directories to search in (8fe277fa5265379affe95e819df8c43969c79d56).
I've also added the optimization for Alldebrid downloads (38a3bc1537be3c76367017d4f787c54e3e6149ff).
@rogerfar The PR (#652) should be ready to merge now if it works for @Cucumberrbob.

**Comment by billouetaudrey** [comment]
Hey since latest version was working and suddenly stop working since yesterday.

Anyone have idea?😘
